File: scripts/type-consistency.py
#!/usr/bin/env python3
"""三方一致性检查。对比 resource_types.json ↔ Go 常量 ↔ JS 常量。"""
import argparse
import json
import sys
import re
from pathlib import Path

ROOT = Path(__file__).resolve().parent.parent


def read_resource_types():
    fp = ROOT / "resource_types.json"
    data = json.loads(fp.read_text("utf-8"))
    types = {}
    for rt in data["resourceTypes"]:
        types[rt["id"]] = rt
    return types


def read_js_extensions():
    fp = ROOT / "frontend/js/utils/extensions.js"
    text = fp.read_text("utf-8", errors="replace")
    # 提取 RESOURCE_EXTS 对象
    m = re.search(r"export const RESOURCE_EXTS = \{([^}]+)\}", text, re.DOTALL)
    if not m:
        return {}
    body = m.group(1)
    types = {}
    for line in body.split("\n"):
        line = line.strip().rstrip(",")
        if not line or line.startswith("//"):
            continue
        # "key": [".ext1", ".ext2"]
        m2 = re.match(r'"?(\w[\w-]*)"?\s*:\s*\[([^\]]+)\]', line)
        if m2:
            key = m2.group(1)
            exts = re.findall(r'"([^"]+)"', m2.group(2))
            types[key] = exts
    return types


def read_go_constants():
    """从 Go 源码提取静态 map（仅限 hardcoded fallback）。"""
    fp = ROOT / "go/types/extensions.go"
    text = fp.read_text("utf-8", errors="replace")
    return text  # 目前 Go 端全动态，没有静态 map


def main():
    parser = argparse.ArgumentParser()
    parser.add_argument("--json", action="store_true")
    args = parser.parse_args()

    json_types = read_resource_types()
    js_types = read_js_extensions()

    issues = []

    # JSON → JS: 检查 JS 是否缺失类型
    for tid, rt in json_types.items():
        if tid not in js_types:
            issues.append({
                "type": "missing_in_js", "id": tid,
                "json_exts": rt["extensions"],
                "detail": f"resource_types.json 有 {tid}，但 extensions.js 没有"
            })
        else:
            js_exts = js_types[tid]
            json_exts = rt["extensions"]
            if set(js_exts) != set(json_exts):
                issues.append({
                    "type": "ext_mismatch", "id": tid,
                    "json_exts": json_exts, "js_exts": js_exts,
                    "detail": f"{tid}: JSON={json_exts} JS={js_exts}"
                })

    # JS → JSON: 检查 JS 是否有多余的类型
    for tid in js_types:
        if tid not in json_types:
            issues.append({
                "type": "extra_in_js", "id": tid,
                "js_exts": js_types[tid],
                "detail": f"extensions.js 有 {tid}，但 resource_types.json 没有"
            })

    if args.json:
        out = {"_summary": {"issues": len(issues)}, "issues": issues}
        sys.stdout.buffer.write(json.dumps(out, ensure_ascii=False, indent=2).encode("utf-8"))
        sys.stdout.buffer.write(b"\n")
    else:
        if issues:
            for i in issues:
                sys.stdout.buffer.write(f"[{i['type']}] {i['detail']}\n".encode("utf-8"))
        else:
            sys.stdout.buffer.write("全部一致\n".encode("utf-8"))

if __name__ == "__main__":
    main()

File: scripts/type-consistency.ts
#!/usr/bin/env node
/**
 * type-consistency.ts — resource_types.json 单一事实来源「派生守卫」。
 *
 * 设计意图：守护 ADR-014 / T2「extensions.ts 必须派生自 resource_types.json，
 * 禁止手写 RESOURCE_EXTS 字面量副本」这一单一事实来源红线。
 *
 * 历史演进（ADR-204）：
 *   本脚本原做 resource_types.json ↔ extensions.ts 字面量比对（JSON↔JS）。
 *   T2 后 extensions.ts 改为构建期从 schema.ts / resource_types.json 派生
 *   （Object.fromEntries），字面量比对随之成为**不可达死代码**——readJsExtensions
 *   派生完好即返回 null（比对分支永不进），派生被破坏即 throw(fatal)。
 *   比对职责已下沉 Go 侧 resource_types_consistency_test.go（dup id 等结构性校验）。
 *   故本脚本收敛为「派生守卫」单一职责：确保 extensions.ts 仍从 JSON 派生，
 *   而非回潮为手写副本（手写副本 tsc 抓不到，只有此正则守卫能拦）。
 *
 * 依赖：node:fs / node:path / _lib/scan-files.ts（getRoot）。
 *
 * 退出码：派生完好 → 0（无 issues）；派生被破坏（手写副本）→ 1（fatal，_summary.issues=9999）。
 *
 * 用法：
 *   node scripts/type-consistency.ts            # 文本报告
 *   node scripts/type-consistency.ts --json     # JSON 输出（CI / 子代理 / pre-push-gate 消费）
 */
import fs from 'node:fs';
import path from 'node:path';
import { getRoot } from './_lib/scan-files.ts';

const ROOT = getRoot();

/**
 * 校验 extensions.ts 仍从 resource_types.json 派生 RESOURCE_EXTS。
 * 派生完好 → 返回（不抛）；手写副本 / 派生链路断裂 → 抛错走 fatal（fail-closed）。
 *
 * 链路判定：
 *   - 必须 `export const RESOURCE_EXTS` 且经 `Object.fromEntries` 派生；
 *   - 入口须为链路 A（直接 import resource_types.json）或链路 B（import allResourceTypes from ./schema.ts，T2 后唯一入口）。
 */
function assertExtensionsDerived() {
  const fp = path.join(ROOT, 'frontend/src/utils/resource/extensions.ts');
  const text = fs.readFileSync(fp, 'utf-8');
  const derives = /export const RESOURCE_EXTS/.test(text) && text.includes('Object.fromEntries');
  const viaDirectImport = /import resourceTypesJson\b[\s\S]*resource_types\.json/.test(text);
  const viaSchema = /import\s+\{[^}]*allResourceTypes[^}]*\}\s+from\s+["']\.\/schema\.ts["']/.test(text);
  if (!derives || (!viaDirectImport && !viaSchema)) {
    throw new Error('extensions.ts 未从 resource_types.json 派生 RESOURCE_EXTS（ADR-014/T2 单一事实来源被破坏，勿手写副本）');
  }
}

const jsonMode = process.argv.slice(2).includes('--json');

function emit(issues: number) {
  if (jsonMode) {
    process.stdout.write(JSON.stringify({ _summary: { issues }, issues: issues ? [{ type: 'fatal', id: '', detail: '派生守卫失败' }] : [] }, null, 2) + '\n');
  } else if (issues) {
    process.stdout.write('派生守卫失败\n');
  } else {
    process.stdout.write('extensions.ts 派生链路完好（单一事实来源守护通过）\n');
  }
}

try {
  assertExtensionsDerived();
  emit(0);
} catch (e: any) {
  // 派生破坏：输出哨兵 JSON，pre-push-gate 的 `?? 0` 读到 9999 → blocked（fail-closed）。
  if (jsonMode) {
    process.stdout.write(JSON.stringify({ _summary: { issues: 9999 }, issues: [{ type: 'fatal', id: '', detail: `派生守卫失败: ${e?.message || e}` }] }, null, 2) + '\n');
  } else {
    process.stderr.write(`FATAL: ${e?.message || e}\n`);
  }
  process.exit(1);
}
